File: tape/sdk/typescript/test/dx.test.ts
// Smoke tests for the standalone-DX additions (durable / outbox /
// connectors / obs / tenancy). No tape-server required.

import { test } from 'node:test';
import { strict as assert } from 'node:assert';

import {
  outboxTool, isOutboxEnvelope, outboxMetaOf, OutboxConfigError,
  ConnectorRegistry, LogConnector,
  logJson, ALL_SPANS, SPAN_BEGIN_EFFECT,
  tenancyDefaults, tenancyFromObject, warnIfHardButUnenforced,
} from '../src/index.ts';

test('outboxTool rejects unsafe non_idempotent', () => {
  assert.throws(
    () => outboxTool(
      ({ x }: { x: string }) => ({ x }),
      { name: 'wire', connector: 'bank.wire', semantics: 'non_idempotent' },
    ),
    OutboxConfigError,
  );
});

test('outboxTool produces envelope + business_key', () => {
  const wire = outboxTool(
    ({ account, amount, beneficiary }: { account: string; amount: number; beneficiary: string }) =>
      ({ account, amount, beneficiary }),
    {
      name: 'wire_money',
      connector: 'bank.wire',
      semantics: 'non_idempotent',
      businessKey: (p) => `${p.account}:${p.amount}`,
      waitForResult: true,
    },
  );
  const env = wire({ account: 'ACME-1', amount: 100, beneficiary: 'bob' });
  assert.equal(env.__outbox__, true);
  assert.equal(env.connector, 'bank.wire');
  assert.equal(env.tool, 'wire_money');
  assert.equal(env.business_key, 'ACME-1:100');
  assert.equal(env.wait_for_result, true);
  assert.deepEqual(env.payload, { account: 'ACME-1', amount: 100, beneficiary: 'bob' });
  assert.ok(isOutboxEnvelope(env));
  const meta = outboxMetaOf(wire);
  assert.equal(meta?.connector, 'bank.wire');
  assert.equal(meta?.semantics, 'non_idempotent');
});

test('outboxTool rejects async + non-object returns', () => {
  // Non-object return is enforced at call time.
  const bogus = outboxTool((_: any) => 'not-an-object' as any, {
    name: 'b', connector: 'c', semantics: 'idempotent',
  });
  assert.throws(() => bogus({}), OutboxConfigError);
});

test('ConnectorRegistry round-trips', async () => {
  const r = new ConnectorRegistry();
  const c = new LogConnector('/tmp/tape-ts-test.jsonl');
  r.register('log', c);
  assert.throws(() => r.register('log', c), /already registered/);
  assert.equal(r.get('log'), c);
  assert.throws(() => r.get('missing'), /unknown connector/);
  const res = await c.dispatch({
    runId: 'r1', idempotencyKey: 'k1', toolName: 't',
    connector: 'log', payload: { x: 1 },
  });
  assert.equal(res.outcome, 'confirmed');
});

test('logJson emits JSON with canonical ordering', () => {
  // Capture stderr.
  const chunks: string[] = [];
  const orig = process.stderr.write.bind(process.stderr);
  (process.stderr as any).write = (chunk: any) => { chunks.push(String(chunk)); return true; };
  try {
    logJson('hello', { run_id: 'r1', app_name: 'a', reactor: 'recovery', extra: 'x' });
  } finally {
    process.stderr.write = orig;
  }
  const line = chunks.join('');
  const parsed = JSON.parse(line.trim());
  assert.equal(parsed.run_id, 'r1');
  assert.equal(parsed.app_name, 'a');
  assert.equal(parsed.reactor, 'recovery');
  assert.equal(parsed.extra, 'x');
});

test('span constants', () => {
  assert.ok(ALL_SPANS.includes(SPAN_BEGIN_EFFECT));
  assert.equal(ALL_SPANS.length, 11);
});

test('tenancy warns on hard mode', () => {
  const t = tenancyFromObject({ mode: 'hard_multi_tenant', tenantId: 'x' });
  assert.equal(warnIfHardButUnenforced(t).length, 1);
  assert.equal(warnIfHardButUnenforced(tenancyDefaults()).length, 0);
});

File: tape/sdk/java/src/test/java/dev/tape/DxTest.java
package dev.tape;

import dev.tape.connectors.Connector;
import dev.tape.connectors.ConnectorRegistry;
import dev.tape.connectors.LogConnector;
import org.junit.jupiter.api.Test;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/** Smoke tests for the standalone-DX additions (no tape-server required). */
class DxTest {

    @Test
    void outboxRejectsUnsafeNonIdempotent() {
        OutboxTool.OutboxConfigError ex = assertThrows(
            OutboxTool.OutboxConfigError.class,
            () -> OutboxTool.builder("wire", "bank.wire")
                .semantics(OutboxTool.Semantics.NON_IDEMPOTENT)
                .build());
        assertTrue(ex.getMessage().contains("non_idempotent"));
    }

    @Test
    void outboxBuildsEnvelopeWithBusinessKey() {
        OutboxTool wire = OutboxTool.builder("wire_money", "bank.wire")
            .semantics(OutboxTool.Semantics.NON_IDEMPOTENT)
            .businessKey(p -> p.get("account") + ":" + p.get("amount"))
            .waitForResult(true)
            .build();
        Map<String, Object> env = wire.envelope(Map.of(
            "account", "ACME-1", "amount", 100, "beneficiary", "bob"));
        assertEquals(Boolean.TRUE, env.get("__outbox__"));
        assertEquals("bank.wire", env.get("connector"));
        assertEquals("wire_money", env.get("tool"));
        assertEquals("ACME-1:100", env.get("business_key"));
        assertEquals(true, env.get("wait_for_result"));
        assertNotNull(env.get("payload"));
        assertTrue(OutboxTool.isEnvelope(env));
    }

    @Test
    void connectorRegistryRoundTrip() throws Exception {
        ConnectorRegistry r = new ConnectorRegistry();
        Connector c = new LogConnector("/tmp/tape-java-test.jsonl");
        r.register("log", c);
        assertThrows(IllegalStateException.class, () -> r.register("log", c));
        assertSame(c, r.get("log"));
        assertThrows(IllegalArgumentException.class, () -> r.get("missing"));
        Connector.Effect e = new Connector.Effect()
            .runId("r1").idempotencyKey("k1").toolName("t")
            .connector("log").payload(Map.of("x", 1));
        Connector.Result res = c.dispatch(e);
        assertEquals(Connector.DispatchOutcome.CONFIRMED, res.outcome);
    }

    @Test
    void tenancyWarnsOnHardMode() {
        Tenancy.Config t = new Tenancy.Config(Tenancy.Mode.HARD_MULTI_TENANT, "x");
        assertEquals(1, t.warnIfHardButUnenforced().size());
        assertEquals(0, Tenancy.defaults().warnIfHardButUnenforced().size());
    }

    @Test
    void obsLogJsonEmitsJsonLine() {
        // Pipe stderr into a tempfile and check the JSON shape.
        Map<String, Object> fields = Map.of(
            "run_id", "r-1", "app_name", "treasury", "reactor", "recovery");
        Obs.logJson("hello", fields);
        // Best-effort: just confirm the call doesn't throw + ALL_SPANS is populated.
        assertEquals(11, Obs.ALL_SPANS.size());
        assertTrue(Obs.ALL_SPANS.contains(Obs.SPAN_BEGIN_EFFECT));
    }

    @Test
    void durableAppRequiresName() {
        assertThrows(IllegalArgumentException.class,
            () -> DurableApp.wire(new DurableApp.Config()));
    }

    @Test
    void logConnectorWritesJsonLine() throws Exception {
        Path tmp = Files.createTempFile("tape-java-dx", ".jsonl");
        LogConnector c = new LogConnector(tmp.toString());
        Connector.Effect e = new Connector.Effect()
            .runId("r1").idempotencyKey("k1").toolName("t")
            .connector("log").payload(Map.of("x", 1));
        c.dispatch(e);
        String contents = Files.readString(tmp);
        assertTrue(contents.contains("\"kind\":\"dispatch\""), contents);
        assertTrue(contents.contains("\"ts_ms\""), contents);
        Files.deleteIfExists(tmp);
    }
}
